packages/mobile/App/services/sync/utils/pullRecordsInBatches.ts

```diff
@@ -1,41 +1,57 @@
 import { calculatePageLimit } from './calculatePageLimit';
 import { SYNC_SESSION_DIRECTION } from '../constants';
 import { PullParams } from '../MobileSyncManager';
+import { type SyncRecord } from '../types';
+
+type PulledPage = { records: SyncRecord[]; pullTime: number };
 
 export const pullRecordsInBatches = async (
-  {
-    centralServer,
-    sessionId,
-    recordTotal,
-    progressCallback = () => {},
-    syncSettings,
-  }: PullParams,
-  processRecords: (records: any) => Promise<void>,
+  { centralServer, syncSettings, sessionId, recordTotal, progressCallback = () => {} }: PullParams,
+  processRecords: (records: SyncRecord[]) => Promise<void>,
 ) => {
-  const { dynamicLimiter: dynamicLimiterSettings } = syncSettings || {};
-  let fromId;
-  let limit = calculatePageLimit(dynamicLimiterSettings);
-  let totalPulled = 0;
+  if (recordTotal === 0) {
+    return;
+  }
 
-  // pull changes a page at a time
-  while (totalPulled < recordTotal) {
+  const fetchPage = async (pageLimit: number, pageFromId?: string): Promise<PulledPage> => {
     const startTime = Date.now();
-    const records = await centralServer.pull(sessionId, limit, fromId);
+    const records = await centralServer.pull(sessionId, pageLimit, pageFromId);
     const pullTime = Date.now() - startTime;
-    const recordsToSave = records.map(r => ({
+    return { records, pullTime };
+  };
+
+  const { dynamicLimiter: dynamicLimiterSettings } = syncSettings || {};
+  let fromId: string | undefined;
+  let limit = calculatePageLimit(syncSettings?.dynamicLimiter);
+  let totalPulled = 0;
+  let current = await fetchPage(limit, fromId);
+
+  while (totalPulled < recordTotal && current.records.length > 0) {
+    // Get next cursor and adjust page size based on how long the last pull took
+    const last = current.records.at(-1);
+    const nextFromId = last
+      ? btoa(JSON.stringify({ sortOrder: last.sortOrder, id: last.id }))
+      : undefined;
+    limit = calculatePageLimit(dynamicLimiterSettings, limit, current.pullTime);
+
+    // Prefetch next page in background
+    const nextPromise = nextFromId
+      ? fetchPage(limit, nextFromId)
+      : Promise.resolve({ records: [], pullTime: 0 });
+
+    // Process current page while next is downloading
+    const recordsToSave = current.records.map(r => ({
       ...r,
       // mark as never updated, so we don't push it back to the central server until the next update
       data: { ...r.data, updated_at_sync_tick: -1 },
       direction: SYNC_SESSION_DIRECTION.INCOMING,
     }));
-
     await processRecords(recordsToSave);
-
-    const { id, sortOrder } = records[records.length - 1];
-    fromId = btoa(JSON.stringify({ sortOrder, id }));
-    totalPulled += records.length;
-    limit = calculatePageLimit(dynamicLimiterSettings, limit, pullTime);
-
+    totalPulled += current.records.length;
     progressCallback(recordsToSave.length);
+
+    // Switch to next page
+    fromId = nextFromId;
```

Comment:
`fromId` is never used for anything meaningful, can delete the variable entirely